### PR TITLE
Check balance in VM create implementations

### PR DESCRIPTION
**Pull request by simonlechner**
LFVM and evmzero did not handle the CREATE instruction correctly if a value higher than the accounts balance has been sent. This is fixed with this PR.
Regression tests are going to be added with the addition of the CREATE instruction to the CT in #441.

### CODE REVIEW

**Review by facuMH — APPROVED** [review]
LGTM !! nice fix

File: cpp/vm/evmzero/interpreter.cc
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter.cc:1]: The changes in this file fixes two different issues:
- the incorrect handling of insufficient balance; it used to be a rollback, but it should be a skipped call
- the incorrect order of the balance check and the memory expansion

For both of those, distinct regression tests should be added to #441

File: go/vm/lfvm/instructions.go [hunk 1]
```diff
@@ -753,6 +754,17 @@ func opCreate(c *context) {
 		return
 	}
 
+	if !value.IsZero() {
+		balance := [32]byte(c.context.GetBalance(c.params.Recipient))
```
Comment (HerbertJordan) [inline go/vm/lfvm/instructions.go:758]: nit: this conversion is not needed
```suggestion
		balance := c.context.GetBalance(c.params.Recipient)
```

File: go/vm/lfvm/instructions.go
Comment (HerbertJordan) [inline go/vm/lfvm/instructions.go:1]: With this, #441 should add 3 regression tests.

File: cpp/vm/evmzero/interpreter.cc
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter.cc:1]: Maybe those changes should also be covered in evmzero unit tests. Wdyt?
For an example on how to do that, see #354.

File: go/vm/lfvm/instructions.go
Comment (HerbertJordan) [inline go/vm/lfvm/instructions.go:1]: Maybe those changes should also be covered in LFVM unit tests. Wdyt?
For an example on how to do that, see #352.

**Review by HerbertJordan — APPROVED** [review]

File: go/vm/lfvm/instructions.go
Comment (simonlechner) [reply go/vm/lfvm/instructions.go:1]: The value check error in lfvm and evmzero is triggered with the same input state, therefor it is sufficient to only add two regression tests.

File: cpp/vm/evmzero/interpreter.cc
Comment (simonlechner) [reply cpp/vm/evmzero/interpreter.cc:1]: Done.

File: go/vm/lfvm/instructions.go
Comment (simonlechner) [reply go/vm/lfvm/instructions.go:1]: Done.

File: cpp/vm/evmzero/interpreter_test.cc [hunk 2]
```diff
@@ -4853,6 +4853,29 @@ TEST(InterpreterTest, SELFDESTRUCT_StaticCallViolation) {
   });
 }
 
+///////////////////////////////////////////////////////////
+// CREATE
+
+TEST(InterpreterTest, CREATE_InsuffiecientBalanceIsHandledCorrectly) {
```
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter_test.cc:4859]: There should be another test case testing that in a case where there is not enough balance the memory has grown to the size of the input data. Before this change, this used to be not the case.

**Review by HerbertJordan — APPROVED** [review]

**Review by HerbertJordan — APPROVED** [review]

File: cpp/vm/evmzero/interpreter_test.cc (hunk 2 above)
Comment (simonlechner) [reply cpp/vm/evmzero/interpreter_test.cc:4859]: Done.

**Pull request merged**